HW_2.py

```diff
@@ -0,0 +1,164 @@
+        self.author = model.author
+        self.year = model.year
+
+    def __str__(self) -> str:
+        return f"'{self._model.book_name}' {self._model.year} {self._model.author}"
+
+    def __repr__(self) -> str:
+        return self.__str__()
+
+    def get_info(self) -> str:
+        return f'Book title: {self._model.book_name}\nAuthor: {self._model.author}\nYear:{self._model.year}'
+
+
+class Magazine(Book):
+    def __init__(self, model: BookModel):
+        super().__init__(model)
+
+    def get_info(self) -> str:
+        return f'Magazine title: {self._model.book_name}\nAuthor: {self._model.author}\nYear:{self._model.year}'
+
+
+class Library:
+    def __init__(self):
+        self.book_collection = []
+
+    def __iter__(self):
+        self.current_book = 0
+        return self
+
+    def __next__(self):
+        if self.current_book >= len(self.book_collection):
+            raise StopIteration
+
+        book = self.book_collection[self.current_book]
+        self.current_book += 1
+        return book
+
+    def book_by_author_gen(self, author: str):
+        return (book for book in self.book_collection if book.author == author)
+
+    def save_data(self):
+        with file_opener('library_list.txt', 'w') as file:
+            for book in self.book_collection:
+                file.write(f'{str(book)}\n')
+
+    def load_data(self):
+        with file_opener('library_list.txt', 'r') as file:
+            return file.read().strip().split('\n')
+
+    def logger(func):
+        def wrapper(*args, **kwargs):
+            result = func(*args, **kwargs)
+            with open('log.txt', 'w', encoding='UTF-8') as file:
+                file.write(f"Книга '{args[1].book_name}' була додана до бібліотеки.\n")
+            return result
+        return wrapper
+
+    def is_exist(func):
+        def wrapper(self, book):
+            if book in self.book_collection:
+                return func(self, book)
+            else:
+                print(f"Книги '{book.book_name}' не існує у бібліотеці")
+        return wrapper
+
+    @logger
+    def add_book(self, book: Book):
+        self.book_collection.append(book)
+
+
+    def add_books(self, books: [Book]):
+        for book in books:
+            self.book_collection.append(book)
+
+    @is_exist
+    def del_book(self, book: Book):
+        if book in self.book_collection:
+            self.book_collection.remove(book)
+
+
+# books_list = Library([book_1, book_2, book_3])
+#
+#
+# not_existing_book_model = BookModel(book_name = 'Basdasad4', author = 'Unknown', year = 165436)
+#
+# books_list.add_book(new_book_model)
+# books_list.add_book(mag_1)
+#
+# books_list.del_book(not_existing_book_model)
+#
+# books_list.save_data()
+# books_list.load_data()
+
+# print(*books_list, sep=',')
+
+# Main code
+
+
+
+Lib = Library() #creating a Library
+
+# Creating models and inctances
+book_model_1 = BookModel(book_name = 'Book1', author = 'Conan Doyle', year = 1256)
```

Comment:
By pep-8 there shouldn't be a spaces in arguments passing. Same for other instances below
```suggestion
book_model_1 = BookModel(book_name='Book1', author='Conan Doyle', year=1256)
```